File: backend/app/calculation/recommender.py
"""価格探索・複数目的推奨ロジック。単一の「適正価格」ではなく、目的別に評価する。"""
from __future__ import annotations

from dataclasses import dataclass

from app.calculation import constants as c
from app.calculation.simulator import PerformanceSimulator
from app.calculation.types import DemandFeatures, ScenarioResult, VenueCandidate


@dataclass
class Recommendation:
    sellout_price: int
    revenue_price: int
    profit_price: int
    balance_price: int
    recommended_price_range: tuple[int, int]
    sellout_scenario: ScenarioResult
    revenue_scenario: ScenarioResult
    profit_scenario: ScenarioResult
    balance_scenario: ScenarioResult
    all_scenarios: list[ScenarioResult]


def _price_range(price_min: int, price_max: int, step: int = c.PRICE_STEP) -> list[int]:
    if price_min > price_max:
        price_min, price_max = price_max, price_min
    prices = list(range(price_min, price_max + 1, step))
    if not prices or prices[-1] != price_max:
        prices.append(price_max)
    return sorted(set(prices))


def run_full_search(
    simulator: PerformanceSimulator,
    features: DemandFeatures,
    venues: list[VenueCandidate],
    num_performances_candidates: list[int],
    price_min: int,
    price_max: int,
) -> list[ScenarioResult]:
    prices = _price_range(price_min, price_max)
    scenarios: list[ScenarioResult] = []
    for venue in venues:
        for num_performances in num_performances_candidates:
            for price in prices:
                scenarios.append(
                    simulator.simulate(features, venue, price, num_performances)
                )
    return scenarios


def _normalize(values: list[float]) -> list[float]:
    lo, hi = min(values), max(values)
    if hi - lo < 1e-9:
        return [1.0 for _ in values]
    return [(v - lo) / (hi - lo) for v in values]


def _occupancy_closeness(occupancy_rate: float) -> float:
    lo, hi = c.BALANCE_TARGET_OCCUPANCY_RANGE
    if lo <= occupancy_rate <= hi:
        return 1.0
    if occupancy_rate < lo:
        return max(0.0, 1 - (lo - occupancy_rate) / lo)
    return max(0.0, 1 - (occupancy_rate - hi) / (1 - hi))


def recommend(scenarios: list[ScenarioResult], baseline_price: float) -> Recommendation:
    if not scenarios:
        raise ValueError("At least one scenario is required to recommend a price.")

    # 満席重視: 目標稼働率以上の中で最も高い価格。無ければ稼働率最大。
    sellout_candidates = [
        s for s in scenarios if s.occupancy_rate >= c.SELLOUT_TARGET_OCCUPANCY
    ]
    if sellout_candidates:
        sellout_scenario = max(sellout_candidates, key=lambda s: s.price)
    else:
        sellout_scenario = max(scenarios, key=lambda s: s.occupancy_rate)

    revenue_scenario = max(scenarios, key=lambda s: s.revenue)
    profit_scenario = max(scenarios, key=lambda s: s.profit)

    occupancy_closeness = [_occupancy_closeness(s.occupancy_rate) for s in scenarios]
    revenue_norm = _normalize([s.revenue for s in scenarios])
    profit_norm = _normalize([s.profit for s in scenarios])
    discount_penalty = [
        max(0.0, (baseline_price - s.price) / baseline_price) if baseline_price else 0.0
        for s in scenarios
    ]

    scores = [
        c.BALANCE_WEIGHT_OCCUPANCY * occ
        + c.BALANCE_WEIGHT_REVENUE * rev
        + c.BALANCE_WEIGHT_PROFIT * prof
        - c.BALANCE_WEIGHT_DISCOUNT_PENALTY * pen
        for occ, rev, prof, pen in zip(
            occupancy_closeness, revenue_norm, profit_norm, discount_penalty
        )
    ]
    best_idx = max(range(len(scenarios)), key=lambda i: scores[i])
    balance_scenario = scenarios[best_idx]
    max_score = scores[best_idx]

    same_combo_prices = [
        s.price
        for s, score in zip(scenarios, scores)
        if s.venue_name == balance_scenario.venue_name
        and s.num_performances == balance_scenario.num_performances
        and score >= max_score * c.RECOMMENDED_RANGE_SCORE_THRESHOLD
    ]
    recommended_price_range = (min(same_combo_prices), max(same_combo_prices))

    return Recommendation(
        sellout_price=sellout_scenario.price,
        revenue_price=revenue_scenario.price,
        profit_price=profit_scenario.price,
        balance_price=balance_scenario.price,
        recommended_price_range=recommended_price_range,
        sellout_scenario=sellout_scenario,
        revenue_scenario=revenue_scenario,
        profit_scenario=profit_scenario,
        balance_scenario=balance_scenario,
        all_scenarios=scenarios,
    )

File: frontend/src/calculation/recommender.ts
/** rule_v0.1: backend/app/calculation/recommender.py の移植。 */
import * as c from "./constants";
import { simulate } from "./simulator";
import type { DemandFeatures, Recommendation, ScenarioResult } from "./types";
import type { VenueCandidate } from "../types";

function priceRange(priceMin: number, priceMax: number, step: number = c.PRICE_STEP): number[] {
  let lo = priceMin;
  let hi = priceMax;
  if (lo > hi) [lo, hi] = [hi, lo];
  const prices: number[] = [];
  for (let p = lo; p <= hi; p += step) prices.push(p);
  if (prices.length === 0 || prices[prices.length - 1] !== hi) prices.push(hi);
  return Array.from(new Set(prices)).sort((a, b) => a - b);
}

export function runFullSearch(
  features: DemandFeatures,
  venues: VenueCandidate[],
  numPerformancesCandidates: number[],
  priceMin: number,
  priceMax: number
): ScenarioResult[] {
  const prices = priceRange(priceMin, priceMax);
  const scenarios: ScenarioResult[] = [];
  for (const venue of venues) {
    for (const numPerformances of numPerformancesCandidates) {
      for (const price of prices) {
        scenarios.push(simulate(features, venue, price, numPerformances));
      }
    }
  }
  return scenarios;
}

function normalize(values: number[]): number[] {
  const lo = Math.min(...values);
  const hi = Math.max(...values);
  if (hi - lo < 1e-9) return values.map(() => 1.0);
  return values.map((v) => (v - lo) / (hi - lo));
}

function occupancyCloseness(occupancyRate: number): number {
  const [lo, hi] = c.BALANCE_TARGET_OCCUPANCY_RANGE;
  if (occupancyRate >= lo && occupancyRate <= hi) return 1.0;
  if (occupancyRate < lo) return Math.max(0.0, 1 - (lo - occupancyRate) / lo);
  return Math.max(0.0, 1 - (occupancyRate - hi) / (1 - hi));
}

export function recommend(scenarios: ScenarioResult[], baselinePrice: number): Recommendation {
  if (scenarios.length === 0) {
    throw new Error("At least one scenario is required to recommend a price.");
  }

  const selloutCandidates = scenarios.filter(
    (s) => s.occupancy_rate >= c.SELLOUT_TARGET_OCCUPANCY
  );
  const selloutScenario =
    selloutCandidates.length > 0
      ? selloutCandidates.reduce((a, b) => (b.price > a.price ? b : a))
      : scenarios.reduce((a, b) => (b.occupancy_rate > a.occupancy_rate ? b : a));

  const revenueScenario = scenarios.reduce((a, b) => (b.revenue > a.revenue ? b : a));
  const profitScenario = scenarios.reduce((a, b) => (b.profit > a.profit ? b : a));

  const occupancyClosenessArr = scenarios.map((s) => occupancyCloseness(s.occupancy_rate));
  const revenueNorm = normalize(scenarios.map((s) => s.revenue));
  const profitNorm = normalize(scenarios.map((s) => s.profit));
  const discountPenalty = scenarios.map((s) =>
    baselinePrice ? Math.max(0.0, (baselinePrice - s.price) / baselinePrice) : 0.0
  );

  const scores = scenarios.map(
    (_, i) =>
      c.BALANCE_WEIGHT_OCCUPANCY * occupancyClosenessArr[i] +
      c.BALANCE_WEIGHT_REVENUE * revenueNorm[i] +
      c.BALANCE_WEIGHT_PROFIT * profitNorm[i] -
      c.BALANCE_WEIGHT_DISCOUNT_PENALTY * discountPenalty[i]
  );
  let bestIdx = 0;
  for (let i = 1; i < scores.length; i++) if (scores[i] > scores[bestIdx]) bestIdx = i;
  const balanceScenario = scenarios[bestIdx];
  const maxScore = scores[bestIdx];

  const sameComboPrices = scenarios
    .filter(
      (s, i) =>
        s.venue_name === balanceScenario.venue_name &&
        s.num_performances === balanceScenario.num_performances &&
        scores[i] >= maxScore * c.RECOMMENDED_RANGE_SCORE_THRESHOLD
    )
    .map((s) => s.price);
  const recommendedPriceRange: [number, number] = [
    Math.min(...sameComboPrices),
    Math.max(...sameComboPrices),
  ];

  return {
    sellout_price: selloutScenario.price,
    revenue_price: revenueScenario.price,
    profit_price: profitScenario.price,
    balance_price: balanceScenario.price,
    recommended_price_range: recommendedPriceRange,
    sellout_scenario: selloutScenario,
    revenue_scenario: revenueScenario,
    profit_scenario: profitScenario,
    balance_scenario: balanceScenario,
    all_scenarios: scenarios,
  };
}
